jhub_apps/hub_client/hub_client.py

```diff
@@ -17,22 +18,72 @@
 logger = structlog.get_logger(__name__)
 
 
+def requires_user_token(func):
+    """Decorator to apply to methods of HubClient to create user token before
+    the method call and revoke them after the method call finishes.
+    """
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
+        token_id = self._create_token_for_user()
+        try:
+            original_method_return = func(self, *args, **kwargs)
+        except Exception as e:
+            raise e
+        finally:
+            self._revoke_token(token_id=token_id)
+        return original_method_return
+    return wrapper
+
+
 class HubClient:
-    def __init__(self, token=None):
-        self.token = token or JUPYTERHUB_API_TOKEN
+    def __init__(self, username=None):
+        self.username = username
+        self.token = JUPYTERHUB_API_TOKEN
         self.jhub_apps_request_id = None
         self._set_request_id()
 
     def _set_request_id(self):
         contextvars = structlog.contextvars.get_contextvars()
         self.jhub_apps_request_id = contextvars.get("request_id")
 
-    def _headers(self):
+    def _headers(self, token=None):
         return {
-            "Authorization": f"token {self.token}",
+            "Authorization": f"token {token or self.token}",
             "JHUB_APPS_REQUEST_ID": self.jhub_apps_request_id
         }
 
+    def _create_token_for_user(self):
+        assert self.username
+        logger.info("Creating token for user", username=self.username)
+        r = requests.post(
+            API_URL + f"/users/{self.username}/tokens",
+            headers=self._headers(token=JUPYTERHUB_API_TOKEN),
+            json={
+                # Expire in 5 minutes max
+                "expires_in": 60*5
+            }
+        )
+        r.raise_for_status()
+        rjson = r.json()
+        self.token = rjson["token"]
+        return rjson["id"]
+
+    def _revoke_token(self, token_id):
+        assert self.username
+        assert token_id
+        logger.info("Revoking token")
+        r = requests.delete(
+            API_URL + f"/users/{self.username}/tokens/{token_id}",
+            headers=self._headers(token=JUPYTERHUB_API_TOKEN),
+        )
+        r.raise_for_status()
+        logger.info(
+            "Token revoked",
+            status_code=r.status_code,
+            username=self.username, token_id=token_id
```

Comment:
For peace of mind I would not include even a revoked token in logs.

```suggestion
            username=self.username
```

Reply:
accepted, although its just the id of the token, not the token.